In [1]:
from sqlalchemy import create_engine, text

In [2]:
database_name = 'scooters'    

connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"

In [3]:
engine = create_engine(connection_string)

In [4]:
import pandas as pd

In [5]:
query = '''
SELECT *
FROM scooters
WHERE pubdatetime IS NULL OR latitude IS NULL OR longitude IS NULL OR sumdid IS NULL OR  sumdtype IS NULL OR  
chargelevel IS NULL OR  sumdgroup IS NULL OR  costpermin IS NULL OR companyname IS NULL;
'''


In [6]:
with engine.connect() as connection:    
    null_scooters = pd.read_sql(text(query), con = connection)

In [7]:
null_scooters

pubdatetime   latitude  longitude  \
0   2019-07-29 00:28:07  36.149802 -86.807646   
1   2019-07-29 00:33:07  36.149802 -86.807646   
2   2019-07-29 00:38:07  36.149802 -86.807646   
3   2019-07-29 00:43:07  36.149802 -86.807646   
4   2019-07-29 00:48:07  36.149802 -86.807646   
..                  ...        ...        ...   
765 2019-07-21 16:27:47  36.152903 -86.762114   
766 2019-07-21 16:32:47  36.152911 -86.762134   
767 2019-07-21 16:37:47  36.152857 -86.762132   
768 2019-07-21 16:47:47  36.152895 -86.762149   
769 2019-07-21 16:57:48  36.152926 -86.762121   

                                           sumdid sumdtype chargelevel  \
0    Powered-55a7ac02-7e17-bb3b-130d-b160d9776405  Powered        None   
1    Powered-55a7ac02-7e17-bb3b-130d-b160d9776405  Powered        None   
2    Powered-55a7ac02-7e17-bb3b-130d-b160d9776405  Powered        None   
3    Powered-55a7ac02-7e17-bb3b-130d-b160d9776405  Powered        None   
4    Powered-55a7ac02-7e17-bb3b-130d-b160d9776405  Powered        None   
..                                            ...      ...         ...   
765  Powered-78f12faa-60fb-82e0-b356-6c7ea446345a  Powered        None   
766  Powered-78f12faa-60fb-82e0-b356-6c7ea446345a  Powered        None   
767  Powered-78f12faa-60fb-82e0-b356-6c7ea446345a  Powered        None   
768  Powered-78f12faa-60fb-82e0-b356-6c7ea446345a  Powered        None   
769  Powered-78f12faa-60fb-82e0-b356-6c7ea446345a  Powered        None   

    sumdgroup  costpermin companyname  
0     Scooter         0.3        Bolt  
1     Scooter         0.3        Bolt  
2     Scooter         0.3        Bolt  
3     Scooter         0.3        Bolt  
4     Scooter         0.3        Bolt  
..        ...         ...         ...  
765   Scooter         0.3        Bolt  
766   Scooter         0.3        Bolt  
767   Scooter         0.3        Bolt  
768   Scooter         0.3        Bolt  
769   Scooter         0.3        Bolt  

[770 rows x 9 columns]

Are there any null values in any columns in either table? YES- 770 nulls found 

What date range is represented in each of the date columns? Investigate any values that seem odd. min date- 2019-05-01 00:01:41.247 : max date- 2019-07-31 23:59:57; min date possibly accoundated for milliseconds (.247) where as max date has %Y-%m-%d-%H:%M:%S format. 

In [20]:
query = '''
SELECT MIN(pubdatetime) AS min_date, MAX(pubdatetime) AS max_date
FROM scooters
'''


In [21]:
with engine.connect() as connection:    
   dates_range = pd.read_sql(text(query), con = connection)

In [24]:
dates_range

min_date            max_date
0 2019-05-01 00:01:41.247 2019-07-31 23:59:57

Is time represented with am/pm or using 24 hour values in each of the columns that include time? 24 hours 

In [ ]:
query = '''
SELECT DISTINCT pubdatetime
FROM scooters
'''
with engine.connect() as connection:    
    time = pd.read_sql(text(query), con = connection)

In [ ]:
time

What values are there in the sumdgroup column? scooter, Scooter and bicycle. 
Are there any that are not of interest for this project? bicycle

In [26]:
query = '''
SELECT DISTINCT sumdgroup
FROM scooters
'''
with engine.connect() as connection:    
    sumdgroup = pd.read_sql(text(query), con = connection)

In [25]:
sumdgroup

NameError: name 'sumdgroup' is not defined

What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [ ]:
query = '''
SELECT MIN(latitude) AS min_latitude, MAX(latitude) AS max_latitude
FROM scooters
'''
with engine.connect() as connection:    
   lat = pd.read_sql(text(query), con = connection)

In [ ]:
lat

In [ ]:
query = '''
SELECT MIN(longitude) AS min_longitude, MAX(longitude) AS max_longitude
FROM scooters
'''
with engine.connect() as connection:    
   lon = pd.read_sql(text(query), con = connection)

In [ ]:
lon

What is the range of values for trip duration and trip distance? Do these values make sense? 
Explore values that might seem questionable.

In [14]:
query = '''
SELECT MIN(tripduration) AS trip_duration_min, MAX(tripduration) AS trip_duration_max , MIN(tripdistance) AS trip_distance_min,
MAX(tripdistance) AS trip_distance_max
FROM trips
'''
with engine.connect() as connection:    
    time_distance = pd.read_sql(text(query), con = connection)

In [15]:
time_distance

trip_duration_min  trip_duration_max  trip_distance_min  trip_distance_max
0         -19.358267           512619.0        -20324803.8       3.188448e+07

In [12]:
query = '''
SELECT *
FROM trips
 ;
'''
with engine.connect() as connection:    
    min_duration = pd.read_sql(text(query), con = connection)

In [13]:
  min_duration

Empty DataFrame
Columns: [pubtimestamp, companyname, triprecordnum, sumdid, tripduration, tripdistance, startdate, starttime, enddate, endtime, startlatitude, startlongitude, endlatitude, endlongitude, triproute, create_dt]
Index: []

Check out how the values for the company name column in the scooters table compare to those of the trips table. 
What do you notice? the caplization is differes and Bolt is called Bolt Mobility in trips table 


In [27]:
query = '''
SELECT DISTINCT companyname
FROM scooters
'''
with engine.connect() as connection:    
    scooters_company_name = pd.read_sql(text(query), con = connection)

In [28]:
scooters_company_name

companyname
0        Bird
1        Bolt
2      Gotcha
3        Jump
4        Lime
5        Lyft
6        Spin

In [29]:
query = '''
SELECT DISTINCT companyname
FROM trips
'''
with engine.connect() as connection:    
    trips_company_name = pd.read_sql(text(query), con = connection)

In [30]:
 trips_company_name

companyname
0           Bird
1  Bolt Mobility
2         Gotcha
3           JUMP
4           Lime
5           Lyft
6           SPIN